Fix support for extras in micropip

amitsaha opened the pull request:
### Description

Fix for https://github.com/pyodide/pyodide/issues/2535

### Checklists

- [x] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry
- [x] Add / update tests

amitsaha commented on packages/micropip/src/micropip/_micropip.py:
```diff
@@ -263,15 +264,24 @@ async def add_requirement_inner(
-            if not req.marker.evaluate(self.ctx):
+            def eval_marker(e: dict[str, str]) -> bool:
+                self.ctx.update(e)
+                # mypy stuff
```
I was getting a mypy error without this additional bool check, will revisit.

amitsaha commented on packages/micropip/test_micropip.py:
```diff
@@ -281,6 +282,23 @@ def test_add_requirement_marker(mock_importlib):
+    asyncio.get_event_loop().run_until_complete(
+        transaction.gather_requirements(
+            [
+                "hypothesis[cli]",
```
Ideally, i would like to update the `pyodide-micropip-test` package here updated with some extras.

ryanking13 commented:
Thanks, @amitsaha. I'll review this before long.

Could you also add test cases of which we can ensure that extra flags don't propagate?

For example, 

```
await micropip.install(["A[all]", "B"])
```

should not install extra dependencies of `B`.

```
await micropip.install("A[all]")
await micropip.install("B")
```

also should not install extra expendencies of `B`.

```
# Package A has
# [options.extras_require]
# all = B
# Package B has
# [options.extras_require]
# all = C
```

then, `await micropip.install("A[all]") should not install C.

Note that you can use `mock_get_pypi_json` and `mock_fetch_bytes` to mock package metadata.

hoodmane commented:
I will make a PR with some improvements to `mock_fetch_bytes` which should make writing these tests easier.

hoodmane commented:
Opened #2589 which should help here.

amitsaha commented:
> I will make a PR with some improvements to `mock_fetch_bytes` which should make writing these tests easier.

This was great, thanks @hoodmane !

amitsaha commented:
Looking at the test failures.

ryanking13 commented:
Thanks, @amitsaha! Feel free to ping us anytime if you need any help.
About the test failure, I've not checked, but maybe we don't need to keep `self.ctx` and `self.ctx_extras` during the transaction.

Maybe we can update:

```py
def eval_marker(e: dict[str, str]) -> bool:
    self.ctx.update(e)
    # need the assertion here to make mypy happy:
    # https://github.com/python/mypy/issues/4805
    assert req.marker is not None
    return req.marker.evaluate(self.ctx)

self.ctx.update({"extra": ""})
if not req.marker.evaluate(self.ctx) and not any(
    [eval_marker(e) for e in self.ctx_extras]
):
    return
```

to something like:

```py
def eval_marker(e: dict[str, str]) -> bool:
    assert req.marker is not None
    return req.marker.evaluate({"extra": e})

if not req.marker.evaluate({"extra": ""}) and not any(
    [eval_marker(e) for e in req.extras]
):
    return
```

amitsaha commented:
@ryanking13 I found the issue of the test failure. The test only fails when you run it with the other test I added: `test_package_with_extra`. 

The reason I think is, in this test, I add `pkga` with `extras` having `opt_feature`, then I add the same package in the failing test with `all` extras. Something in the code doesn't update the extras of `pkga` with the `all` marker. Will look into that next.

hoodmane commented:
We should consider updating the fixtures to actually isolate the different tests (i.e., delete everything in the `WHEEL_BASE` directory on tear down).

amitsaha commented:
> We should consider updating the fixtures to actually isolate the different tests (i.e., delete everything in the `WHEEL_BASE` directory on tear down).

Can you explain your thoughts? Is it because `micropip` already finds the `pkga` installed and hence doesn't bother installing it again and hence doesn't get the new dependencies?

hoodmane commented:
Yes exactly.

hoodmane commented:
One minute, will make PR to fix.

amitsaha commented:
> Yes exactly.

I wanted to see if I could force it to upgrade to a new version, but i get this error:

```
>       raise ValueError(                                                                      
            f"Requested '{req}', " f"but {req.name}=={ver} is already installed"               
        )                                      
E       ValueError: Requested 'pkga[all]==2.0.0', but pkga==1.0.0 is already installed         
                                               
packages/micropip/src/micropip/_micropip.py:253: ValueError     
```

hoodmane commented:
We could support upgrades if you think it would be fun to implement or you see a use case. I think it's pretty low priority.

hoodmane commented:
Opened #2623

amitsaha commented:
The `test-core-firefox` run had a test failure which you folks might find interesting i think:

```
>               raise Flaky(
                    f"Inconsistent test results! Test case was {node.transition!r} "
                    f"on first run but {new_transition!r} on second"
                )
E               hypothesis.errors.Flaky: Inconsistent test results! Test case was Conclusion(status=Status.INTERESTING, interesting_origin=(<class 'selenium.common.exceptions.TimeoutException'>, '/usr/local/lib/python3.10/site-packages/selenium/webdriver/remote/errorhandler.py', 247, (), ())) on first run but Conclusion(status=Status.INTERESTING, interesting_origin=(<class 'pyodide_test_runner.browser.JavascriptException'>, '/root/repo/pyodide-test-runner/pyodide_test_runner/browser.py', 274, (), ())) on second

/usr/local/lib/python3.10/site-packages/hypothesis/internal/conjecture/datatree.py:406: Flaky
---------------------------- Captured stdout setup -----------------------------
Spawning webserver at http://127.0.0.1:53221 (see logs in /tmp/tmpas1ygp_3/http-server.log)
----------------------------- Captured stdout call -----------------------------
JavascriptException message:  Pyodide already fatally failed and can no longer be used.
--------------------------- Captured stdout teardown ---------------------------
Loading pytest, atomicwrites, attrs, six, more-itertools, pluggy, py, setuptools, distutils, pyparsing, iniconfig
Loaded atomicwrites, attrs, pytest, six, more-itertools, py, pluggy, pyparsing, iniconfig, distutils, setuptools
pytest already loaded from default channel
No new packages to load
Python exception:
OverflowError: Python int too large to convert to C ssize_t


Pyodide has suffered a fatal error. Please report this to the Pyodide maintainers.
The cause of the fatal error was:
PythonError: OverflowError: Python int too large to convert to C ssize_t

PythonError@http://127.0.0.1:53221/pyodide.asm.js:10:211756
new_error@http://127.0.0.1:53221/pyodide.asm.js:10:280403
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[342]:0xef483
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[343]:0xef584
_createPyodideModule/Module._pythonexc2js@http://127.0.0.1:53221/pyodide.asm.js:10:975191
get length@http://127.0.0.1:53221/pyodide.asm.js:10:184073
get@http://127.0.0.1:53221/pyodide.asm.js:10:186749
hiwire_has_length@http://127.0.0.1:53221/pyodide.asm.js:10:263870
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[364]:0xefbef
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[365]:0xf0329
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[459]:0xf46e7
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1960]:0x1763a0
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[3173]:0x203aa7
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[3164]:0x200c23
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[3160]:0x1fb242
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1043]:0x136cb5
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1040]:0x136ad5
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1041]:0x136bcf
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1042]:0x136c72
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[3164]:0x201ad8
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1370]:0x14a17e
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1391]:0x14b018
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[810]:0x12355b
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[3164]:0x1fdd00
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1370]:0x14a17e
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1391]:0x14b018
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[810]:0x12355b
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[3164]:0x1fdd00
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1370]:0x14a17e
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1391]:0x14b018
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[810]:0x12355b
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[3164]:0x1fdd00
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1370]:0x14a17e
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1362]:0x149883
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1392]:0x14b026
method_call_trampoline@http://127.0.0.1:53221/pyodide.asm.js:10:280317
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1182]:0x13eeb8
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[3173]:0x203aa7
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[3164]:0x2014e6
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[3160]:0x1fb242
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1043]:0x136cb5
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1086]:0x13866a
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[3336]:0x219624
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1960]:0x1763a0
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1040]:0x136ad5
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1041]:0x136bcf
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1042]:0x136c72
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[3164]:0x201ad8
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[3160]:0x1fb242
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1043]:0x136cb5
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[3173]:0x203aa7
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[3164]:0x2014e6
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[3160]:0x1fb242
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[1043]:0x136cb5
@http://127.0.0.1:53221/pyodide.asm.wasm:wasm-function[434]:0xf3044
_createPyodideModule/</Module.callPyObjectKwargs@http://127.0.0.1:53221/pyodide.asm.js:10:181792
_createPyodideModule/</Module.callPyObject@http://127.0.0.1:53221/pyodide.asm.js:10:182173
wrapper@http://127.0.0.1:53221/pyodide.asm.js:10:247144

Stack (most recent call first):
  File "/root/repo/pyodide-test-runner/pyodide_test_runner/tests/test_decorator.py", line 258 in test_hypothesis
  File "<exec>", line 11 in __tmp
  File "<exec>", line 21 in <module>
  File "/lib/python3.10/site-packages/_pyodide/_base.py", line 355 in run_async
  File "/lib/python3.10/site-packages/_pyodide/_base.py", line 502 in eval_code_async
  File "/lib/python3.10/asyncio/tasks.py", line 232 in __step
  File "/lib/python3.10/asyncio/events.py", line 80 in _run
  File "/lib/python3.10/site-packages/pyodide/webloop.py", line 157 in run_handle
=============================== warnings summary ===============================
../../usr/local/lib/python3.10/site-packages/pytest_asyncio/plugin.py:191
  /usr/local/lib/python3.10/site-packages/pytest_asyncio/plugin.py:191: DeprecationWarning: The 'asyncio_mode' default value will change to 'strict' in future, please explicitly use 'asyncio_mode=strict' or 'asyncio_mode=auto' in pytest configuration file.
    config.issue_config_time_warning(LEGACY_MODE, stacklevel=2)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
------------ generated xml file: /root/repo/test-results/junit.xml -------------
============================= slowest 50 durations =============================
22.08s call     pyodide-test-runner/pyodide_test_runner/tests/test_decorator.py::test_hypothesis[firefox]
4.29s setup    pyodide-test-runner/pyodide_test_runner/tests/test_decorator.py::test_hypothesis[firefox]
0.92s teardown pyodide-test-runner/pyodide_test_runner/tests/test_decorator.py::test_hypothesis[firefox]
=========================== short test summary info ============================
FAILED pyodide-test-runner/pyodide_test_runner/tests/test_decorator.py::hypothesis[firefox]
======================== 1 failed, 1 warning in 31.92s =========================

Exited with code exit status 1

CircleCI received exit code 1
```

@ryanking13 @hoodmane @rth

amitsaha commented:
> Thanks, @amitsaha! Feel free to ping us anytime if you need any help. About the test failure, I've not checked, but maybe we don't need to keep `self.ctx` and `self.ctx_extras` during the transaction.

Thanks @ryanking13. Please see my reply below:

> 
> Maybe we can update:
> 
> ```python
> def eval_marker(e: dict[str, str]) -> bool:
>     self.ctx.update(e)
>     # need the assertion here to make mypy happy:
>     # https://github.com/python/mypy/issues/4805
>     assert req.marker is not None
>     return req.marker.evaluate(self.ctx)
> 
> self.ctx.update({"extra": ""})
> if not req.marker.evaluate(self.ctx) and not any(
>     [eval_marker(e) for e in self.ctx_extras]
> ):
>     return
> ```
> 
> to something like:
> 
> ```python
> def eval_marker(e: dict[str, str]) -> bool:
>     assert req.marker is not None
>     return req.marker.evaluate({"extra": e})
> 
> if not req.marker.evaluate({"extra": ""}) and not any(
>     [eval_marker(e) for e in req.extras]
> ):
>     return
> ```

From what I found the `req.extras` is only set for the package which we are going to install, and not the dependencies. Hence, I thought I will add a class attribute, `ctx_extras` to do so. Hence, if we update the code as per your suggestion, the evaluation will return `False` for the dependencies.

Regarding using `req.marker.evaluate({"extra": e})` instead of updating `self.ctx` with the `extra` dict inside the loop, my reasoning for doing the latter was that will ensure that the other environment variables will continue to be available along with the `extra` dict.

hoodmane commented:
> The [test-core-firefox run](https://circleci.com/gh/pyodide/pyodide/47322?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) had a test failure which you folks might find interesting i think:

Yeah, it is quite interesting. I can't reproduce it but I can at least find a related bug:
```py
>>> a = run_js("({length : 1e+25})"); len(a) 
SystemError: <built-in function len> returned NULL without setting an exception
>>> a = run_js("({length : 1e+26})"); len(a) 
0
```

hoodmane reviewed (approved):
This is a small patch but it is quite confusing. I added some of my own tests, it seems to work correctly. If you understand the logic, it might be helpful to add a detailed description in the comments.

What the heck is `ctx`? What is it for? What is `ctx_extras`? What is it for? I would have to think more carefully to understand, ideally would be nice to have an explanation in the comments.

Other than that, it looks great. Thanks a lot @amitsaha! I'm going to push to this branch and add another test.

hoodmane commented:
Could you add a changelog entry @amitsaha? Then it will be ready to merge.

amitsaha commented:
> This is a small patch but it is quite confusing. I added some of my own tests, it seems to work correctly. If you understand the logic, it might be helpful to add a detailed description in the comments.
> 
> What the heck is `ctx`? What is it for? What is `ctx_extras`? What is it for? I would have to think more carefully to understand, ideally would be nice to have an explanation in the comments.

Completely agree. I have added some comments in the update. I have also written a more detailed explanation of my thought process and understanding here: https://gist.github.com/amitsaha/df2141e7e798d946ae559bf34e74ab42 - hope that helps, and feel free to ask any followup questions.

hoodmane commented:
Okay your explanation makes sense. I'm still confused, but the example helps a lot. Thanks again for working this out @amitsaha!

hoodmane commented:
So is what we'd want to do just to delete the "extras" part of the marker entirely?

hoodmane commented:
I guess not.

hoodmane commented:
I will merge this then.

hoodmane commented:
Thanks again @amitsaha!

ryanking13 commented:
Thanks for your work and for the detailed explanation @amitsaha!
